### ISSUE TITLE
The boomerang causes a potential heap buffer overflow when it hits an enemy

### ISSUE THREAD

**Issue by ntfwc** [ZQuestClassic/ZQuestClassic#131]
AddressSanitizer flags a heap-buffer-overflow when the boomerang hits an enemy at line 1229 of guys.cpp.

Results from a Valgrind session when hitting octoroks with the boomerang:
```
==30019== Conditional jump or move depends on uninitialised value(s)
==30019==    at 0x815441F: enemy::resolveEnemyDefence(weapon*) (guys.cpp:886)
==30019==    by 0x8154DD7: enemy::defenditemclass(int, int*, weapon*) (guys.cpp:1483)
==30019==    by 0x81559E9: enemy::takehit(weapon*) (guys.cpp:1790)
==30019==    by 0x816A57E: eStalfos::takehit(weapon*) (guys.cpp:7442)
==30019==    by 0x81837F8: check_collisions() (guys.cpp:14306)
==30019==    by 0x807E1F0: game_loop() (zelda.cpp:2388)
==30019==    by 0x8080A6A: main (zelda.cpp:3305)
==30019== 
==30019== Use of uninitialised value of size 4
==30019==    at 0x81549CF: enemy::defend(int, int*, int) (guys.cpp:1229)
==30019==    by 0x8154DE6: enemy::defenditemclass(int, int*, weapon*) (guys.cpp:1483)
==30019==    by 0x81559E9: enemy::takehit(weapon*) (guys.cpp:1790)
==30019==    by 0x816A57E: eStalfos::takehit(weapon*) (guys.cpp:7442)
==30019==    by 0x81837F8: check_collisions() (guys.cpp:14306)
==30019==    by 0x807E1F0: game_loop() (zelda.cpp:2388)
==30019==    by 0x8080A6A: main (zelda.cpp:3305)
==30019== 
==30019== Conditional jump or move depends on uninitialised value(s)
==30019==    at 0x8154A69: enemy::defend(int, int*, int) (guys.cpp:1262)
==30019==    by 0x8154DE6: enemy::defenditemclass(int, int*, weapon*) (guys.cpp:1483)
==30019==    by 0x81559E9: enemy::takehit(weapon*) (guys.cpp:1790)
==30019==    by 0x816A57E: eStalfos::takehit(weapon*) (guys.cpp:7442)
==30019==    by 0x81837F8: check_collisions() (guys.cpp:14306)
==30019==    by 0x807E1F0: game_loop() (zelda.cpp:2388)
==30019==    by 0x8080A6A: main (zelda.cpp:3305)
==30019== 
```
Operating System: Xubuntu 16.04
Commit: 6d94c91351eaac44574ad4128539323a07cc2c95

**Comment by evouga** [comment]
This is caused by enemy::defense being unitinitalized. The enemy defense code in general looks to be in a half-implemented/broken state. @ZoriaRPG what is going on here?

**Comment by ZoriaRPG** [comment]
I will check this out. Just saw the notification.

@evouga 
Does this occur with any other weapon type?

The defence code is a full rewrite. I did not run into this particular issue, when i was testing it. Insofar as 'half-implemented or broken', I'm not sure what you mean there.  I intentionally left in a bunch of debugging code so that I would not need to re-implement it to test stuff each time. 

I need to re-implement shield checking code, certainly, but that was the only segment that I did not finish, prior to you asking me to stop, to work on your internal changes. 

@ntfwc 
Does this overflow occur if using my original code ( https://github.com/ZoriaRPG/ZeldaClassic/tree/20_Enemy_Editor_Size_eRock ) ?
